kernel: Flush in ChainstateManager destructor

sedited opened the pull request:
This simplifies the shutdown code a bit and reduces the amount of extra logic required in kernel/bitcoinkernel.cpp to tear down the chainman.

Moving the second flush out of the Shutdown function into the ChainstateManager destructor, should be safe since there should be no new events created after the first flush. Tweak the chainman tests to actually exercise this new tear down behaviour.

---

This PR is part of the [libbitcoinkernel project](https://github.com/bitcoin/bitcoin/issues/27587)

sedited committed:
kernel: Flush in ChainstateManager dtor

Moving the second flush out of the Shutdown function into the
ChainstateManager destructor, should be safe since there should be no
new events created after the first flush.

This simplifies the shutdown code a bit and allows for getting rid of
the `goto` in bitcoin-chainstate.

sedited committed:
validation: Remove ResetCoinsViews

Now that its only callsite has been moved to the destructor, let the
smart pointer do its thing.

sedited committed:
tests: Use chainman dtor logic to simulate restart

Also remove the reference to the destructed chainman, which could cause
UB once the chainman is reset.

sedited committed:
kernel: Remove chainman destructor in bitcoinkernel.cpp

There is no need to flush the background callbacks, since the immediate
task runner is used for the validation signals, which does not have any
pending callbacks. This allows for removal of the code in the
destructor.

sedited committed:
shutdown: Tear down mempool after chainman

Next to being more consistent, since tearing down the mempool first
leaves a dangling pointer in the chainman, this also enables the changes
in the next commits, since flushing calls GetCoinsCacheSizeState, which
would access this dangling pointer otherwise.

l0rinc commented on src/bitcoin-chainstate.cpp:
```diff
@@ -64,10 +64,10 @@ int main(int argc, char* argv[])
-    assert(kernel::SanityChecks(kernel_context));
+    if (!kernel::SanityChecks(kernel_context)) {
+        std::cerr << "Failed sanity check.";
+        return 0;
```
nit: this being a console app, it might make sense to return a non-zero error code when `cerr` is used

TheCharlatan replied:
Yeah, done.

l0rinc reviewed (commented):
Concept ACK

Closed https://github.com/bitcoin/bitcoin/pull/31362 in favor of this

TheCharlatan commented:
Seems like I am running into this here: https://github.com/bitcoin/bitcoin/pull/25073#issuecomment-1194417234, maybe time to introduce a replace mempool method like Carl suggested there?

sedited committed:
tests: Add replace mempool method

This is done in preparation for the next commit to also change the
mempool pointer within the chainstate once the old mempool is reset.
Without this the old, dangling pointer will be reused when destructing
the chainstate manager.

stickies-v reviewed (commented):
Concept ACK

Simplifies the shutdown logic, and makes ChainstateManager more self contained. I'll need to think a bit more about how to test or verify the safety of changes such as in 7a37f2008372c021541809d06d9300278fd5a8da.

TheCharlatan commented on src/test/util/setup_common.cpp:
```diff
@@ -303,6 +303,20 @@ void ChainTestingSetup::LoadVerifyActivateChainstate()
+    // instead.
+    m_node.mempool.reset();
+    bilingual_str error;
+    m_node.mempool = std::make_unique<CTxMemPool>(MemPoolOptionsForTest(m_node), error);
+    auto& dummy_chainstate{static_cast<DummyChainState&>(m_node.chainman->ActiveChainstate())};
+    dummy_chainstate.SetMempool(m_node.mempool.get());
```
I could not come up with a satisfactory alternative yet, suggestions welcome!

TheCharlatan commented on src/test/util/setup_common.cpp:
```diff
@@ -303,6 +303,20 @@ void ChainTestingSetup::LoadVerifyActivateChainstate()
+    m_node.mempool = std::make_unique<CTxMemPool>(MemPoolOptionsForTest(m_node), error);
+    auto& dummy_chainstate{static_cast<DummyChainState&>(m_node.chainman->ActiveChainstate())};
+    dummy_chainstate.SetMempool(m_node.mempool.get());
+    Assert(error.empty());
```
Yeah, that makes more sense.

TheCharlatan commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -397,7 +386,11 @@ struct SnapshotTestSetup : TestChain100Setup {
             m_node.chainman.reset();
+            // Process all callbacks referring to the old manager before creating a
+            // new one.
+            m_node.validation_signals->SyncWithValidationInterfaceQueue();
```
The chainman reset may trigger validation interface notifications, so I think it makes sense to sync with the queue afterwards? Unlike the shutdown procedure, we keep the validation interface around beyond the restarts there.

stickies-v replied:
> The chainman reset may trigger validation interface notifications

Ah, right, `ChainStateFlushed` would be called. Yeah, I think with that knowledge doing it afterwards makes more sense. Also, callbacks are processed regardless of `SyncWithValidationInterfaceQueue()` being called, so I suppose this doesn't really change anything wrt callbacks assuming `m_node.chainman` exists.

TheCharlatan commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -369,23 +369,12 @@ struct SnapshotTestSetup : TestChain100Setup {
-
         BOOST_TEST_MESSAGE("Simulating node restart");
         {
-            for (Chainstate* cs : chainman.GetAll()) {
-                LOCK(::cs_main);
-                cs->ForceFlushStateToDisk();
-            }
-            // Process all callbacks referring to the old manager before wiping it.
-            m_node.validation_signals->SyncWithValidationInterfaceQueue();
-            LOCK(::cs_main);
-            chainman.ResetChainstates();
```
This seems fine as is? I'd prefer not to document these deeper behaviours, that the caller can expect to "just work".

stickies-v replied:
Sorry, I meant to to remove the flushing reference since that's now automatically handled by clearing chainman, so I agree with your comment.

> Simulate a restart of the node by clearing the existing ChainstateManager, and unloading the block index.

TheCharlatan replied:
I'll do that next time I have to push.

TheCharlatan commented:
Thank you for the review @stickies-v,

Updated 6c3bc478ca343557d798e2b18b5f99c08079ba9a -> 8e86e77311a2c76fd34b4481f83205b2432b151a ([chainman_flush_destructor_0](https://github.com/TheCharlatan/bitcoin/tree/chainman_flush_destructor_0) -> [chainman_flush_destructor_1](https://github.com/TheCharlatan/bitcoin/tree/chainman_flush_destructor_1), [compare](https://github.com/TheCharlatan/bitcoin/compare/chainman_flush_destructor_0..chainman_flush_destructor_1))

* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/31382#discussion_r1888637633), moved `Assert` checking that there is no mempool construction error to right after constructing the mempool.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/31382#discussion_r1900712774), consolidated teardown notification comment into a single block.

mzumsande reviewed (commented):
Concept ACK

TheCharlatan commented on src/validation.cpp:
```diff
@@ -6285,6 +6285,12 @@ ChainstateManager::ChainstateManager(const util::SignalInterrupt& interrupt, Opt
     LOCK(::cs_main);
+    for (Chainstate* chainstate : GetAll()) {
+        if (chainstate->CanFlushToDisk()) {
+            chainstate->ForceFlushStateToDisk();
```
AFAIU the second flush is the original one, while the second flush was introduced in https://github.com/bitcoin/bitcoin/commit/3192975f1d177aa9f0bbd823c6387cfbfa943610 to help cleanly process all remaining callbacks. Maybe it could have removed the second flush in that commit?

That said I think it is good practice to be defensive here and keep both.

TheCharlatan commented:
Rebased 8e86e77311a2c76fd34b4481f83205b2432b151a -> fe7d09b5017b748fc71dab2a8bb186ff2c1624bb ([chainman_flush_destructor_1](https://github.com/TheCharlatan/bitcoin/tree/chainman_flush_destructor_1) -> [chainman_flush_destructor_2](https://github.com/TheCharlatan/bitcoin/tree/chainman_flush_destructor_2), [compare](https://github.com/TheCharlatan/bitcoin/compare/chainman_flush_destructor_1..chainman_flush_destructor_2))

* Fixed a silent merge conflict

TheCharlatan commented:
Rebased fe7d09b5017b748fc71dab2a8bb186ff2c1624bb -> 1c299f70833188654b9cf3936d63d370856a408d ([chainman_flush_destructor_2](https://github.com/TheCharlatan/bitcoin/tree/chainman_flush_destructor_2) -> [chainman_flush_destructor_3](https://github.com/TheCharlatan/bitcoin/tree/chainman_flush_destructor_3), [compare](https://github.com/TheCharlatan/bitcoin/compare/chainman_flush_destructor_2..chainman_flush_destructor_3))

* Rebased to get the new bitcoin-chainstate functional test

TheCharlatan commented:
Rebased 1c299f70833188654b9cf3936d63d370856a408d -> e219f15976f50e0a703bc696b6445feb9950e65d ([chainman_flush_destructor_3](https://github.com/TheCharlatan/bitcoin/tree/chainman_flush_destructor_3) -> [chainman_flush_destructor_4](https://github.com/TheCharlatan/bitcoin/tree/chainman_flush_destructor_4), [compare](https://github.com/TheCharlatan/bitcoin/compare/chainman_flush_destructor_3..chainman_flush_destructor_4))

* Fixed conflict with #32154

TheCharlatan commented:
Rebased e219f15976f50e0a703bc696b6445feb9950e65d -> eec7c49cf777cdcc53aeb20e31fbd5881c2b008b ([chainman_flush_destructor_4](https://github.com/TheCharlatan/bitcoin/tree/chainman_flush_destructor_4) -> [chainman_flush_destructor_5](https://github.com/TheCharlatan/bitcoin/tree/chainman_flush_destructor_5), [compare](https://github.com/TheCharlatan/bitcoin/compare/chainman_flush_destructor_4..chainman_flush_destructor_5))

* Rebased to get a CI fix for windows cross

davidgumberg commented on src/init.cpp:
```diff
@@ -388,9 +388,9 @@ void Shutdown(NodeContext& node)
     if (node.validation_signals) {
         node.validation_signals->UnregisterAllValidationInterfaces();
     }
+    node.chainman.reset();
```
In commit ef83e9569752b4273e6aed3124642dc057cab540 (shutdown: Tear down mempool after chainman)

Feel free to disregard as out of scope, but while addressing this, it might be nice to make reintroducing this dangling pointer impossible by making the various mempool pointers `shared_ptr`, e.g. https://github.com/davidgumberg/bitcoin/commit/75c839b8d77c207a0946294672b703d9f8eecbe1

TheCharlatan replied:
Thanks for taking a look :)

I think the problem here is that we are moving the mempool pointer around between different objects, while they don't really share ownership of it at any one time. It not having a value at some points in time does have a logical meaning in the current code. In general I don't like shared pointers for a number of reasons. In this instance I agree that it would solve the problem of not knowing when to destruct the mempool. I am a bit apprehensive to commit this change here though, because I feel like the current design with a unique pointer and raw pointers signalling optional pointer values was done on purpose over a shared pointer design.

TheCharlatan commented:
Rebased 8512349ccd73c7acd8ccd6d4c2708459bdce3301 -> 6137cecfd3a448c2befe043849fce0b4ea2cd1fe ([chainman_flush_destructor_6](https://github.com/TheCharlatan/bitcoin/tree/chainman_flush_destructor_6) -> [chainman_flush_destructor_7](https://github.com/TheCharlatan/bitcoin/tree/chainman_flush_destructor_7), [compare](https://github.com/TheCharlatan/bitcoin/compare/chainman_flush_destructor_6..chainman_flush_destructor_7))

TheCharlatan commented:
Rebased d9cc26d3b2b5c6a62937f1848f99ed430ed4bade -> ba29aaf28b6f0f63d2fec546fe1c60fe216d6c3a ([chainman_flush_destructor_8](https://github.com/TheCharlatan/bitcoin/tree/chainman_flush_destructor_8) -> [chainman_flush_destructor_9](https://github.com/TheCharlatan/bitcoin/tree/chainman_flush_destructor_9), [compare](https://github.com/TheCharlatan/bitcoin/compare/chainman_flush_destructor_8..chainman_flush_destructor_9))

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/31382#issuecomment-3576740793

> I don't think this is accurate?

Thanks, updated.

sedited commented:
Closing this, since it has not received review in a year, and it is not a pressing change.
